### DISCUSSION TITLE
Missing key(s) in state_dict how do i fix this problem?

### DISCUSSION THREAD

**Discussion by way790110** [STVIR/pysot#320] [General] [answered]
pysot/pysot-master$ bash demo.sh
Traceback (most recent call last):
  File "tools/demo.py", line 109, in <module>
    main()
  File "tools/demo.py", line 69, in main
    map_location=lambda storage, loc: storage.cpu()))
  File "/home/ivpl/anaconda3/envs/pysot/lib/python3.7/site-packages/torch/nn/modules/module.py", line 719, in load_state_dict
    self.__class__.__name__, "\n\t".join(error_msgs)))
RuntimeError: Error(s) in loading state_dict for ModelBuilder:
	Missing key(s) in state_dict: "backbone.layer0.0.weight", "backbone.layer0.1.weight", "backbone.layer0.1.bias", "backbone.layer0.1.running_mean", "backbone.layer0.1.running_var", "backbone.layer1.0.conv.0.weight", "backbone.layer1.0.conv.1.weight", "backbone.layer1.0.conv.1.bias", "backbone.layer1.0.conv.1.running_mean", "backbone.layer1.0.conv.1.running_var", "backbone.layer1.0.conv.3.weight", "backbone.layer1.0.conv.4.weight", "backbone.layer1.0.conv.4.bias", "backbone.layer1.0.conv.4.running_mean", "backbone.layer1.0.conv.4.running_var", "backbone.layer1.0.conv.6.weight", "backbone.layer1.0.conv.7.weight", "backbone.layer1.0.conv.7.bias", "backbone.layer1.0.conv.7.running_mean", "backbone.layer1.0.conv.7.running_var", "backbone.layer2.0.conv.0.weight", "backbone.layer2.0.conv.1.weight", "backbone.layer2.0.conv.1.bias", "backbone.layer2.0.conv.1.running_mean", "backbone.layer2.0.conv.1.running_var", "backbone.layer2.0.conv.3.weight", "backbone.layer2.0.conv.4.weight", "backbone.layer2.0.conv.4.bias", "backbone.layer2.0.conv.4.running_mean", "backbone.layer2.0.conv.4.running_var", "backbone.layer2.0.conv.6.weight", "backbone.layer2.0.conv.7.weight", "backbone.layer2.0.conv.7.bias", "backbone.layer2.0.conv.7.running_mean", "backbone.layer2.0.conv.7.running_var", "backbone.layer2.1.conv.0.weight", "backbone.layer2.1.conv.1.weight", "backbone.layer2.1.conv.1.bias", "backbone.layer2.1.conv.1.running_mean", "backbone.layer2.1.conv.1.running_var", "backbone.layer2.1.conv.3.weight", "backbone.layer2.1.conv.4.weight", "backbone.layer2.1.conv.4.bias", "backbone.layer2.1.conv.4.running_mean", "backbone.layer2.1.conv.4.running_var", "backbone.layer2.1.conv.6.weight", "backbone.layer2.1.conv.7.weight", "backbone.layer2.1.conv.7.bias", "backbone.layer2.1.conv.7.running_mean", "backbone.layer2.1.conv.7.running_var", "backbone.layer3.0.conv.0.weight", "backbone.layer3.0.conv.1.weight", "backbone.layer3.0.conv.1.bias", "backbone.layer3.0.conv.1.running_mean", "backbone.layer3.0.conv.1.running_var", "backbone.layer3.0.conv.3.weight", "backbone.layer3.0.conv.4.weight", "backbone.layer3.0.conv.4.bias", "backbone.layer3.0.conv.4.running_mean", "backbone.layer3.0.conv.4.running_var", "backbone.layer3.0.conv.6.weight", "backbone.layer3.0.conv.7.weight", "backbone.layer3.0.conv.7.bias", "backbone.layer3.0.conv.7.running_mean", "backbone.layer3.0.conv.7.running_var", "backbone.layer3.1.conv.0.weight", "backbone.layer3.1.conv.1.weight", "backbone.layer3.1.conv.1.bias", "backbone.layer3.1.conv.1.running_mean", "backbone.layer3.1.conv.1.running_var", "backbone.layer3.1.conv.3.weight", "backbone.layer3.1.conv.4.weight", "backbone.layer3.1.conv.4.bias", "backbone.layer3.1.conv.4.running_mean", "backbone.layer3.1.conv.4.running_var", "backbone.layer3.1.conv.6.weight", "backbone.layer3.1.conv.7.weight", "backbone.layer3.1.conv.7.bias", "backbone.layer3.1.conv.7.running_mean", "backbone.layer3.1.conv.7.running_var", "backbone.layer3.2.conv.0.weight", "backbone.layer3.2.conv.1.weight", "backbone.layer3.2.conv.1.bias", "backbone.layer3.2.conv.1.running_mean", "backbone.layer3.2.conv.1.running_var", "backbone.layer3.2.conv.3.weight", "backbone.layer3.2.conv.4.weight", "backbone.layer3.2.conv.4.bias", "backbone.layer3.2.conv.4.running_mean", "backbone.layer3.2.conv.4.running_var", "backbone.layer3.2.conv.6.weight", "backbone.layer3.2.conv.7.weight", "backbone.layer3.2.conv.7.bias", "backbone.layer3.2.conv.7.running_mean", "backbone.layer3.2.conv.7.running_var", "backbone.layer4.0.conv.0.weight", "backbone.layer4.0.conv.1.weight", "backbone.layer4.0.conv.1.bias", "backbone.layer4.0.conv.1.running_mean", "backbone.layer4.0.conv.1.running_var", "backbone.layer4.0.conv.3.weight", "backbone.layer4.0.conv.4.weight", "backbone.layer4.0.conv.4.bias", "backbone.layer4.0.conv.4.running_mean", "backbone.layer4.0.conv.4.running_var", "backbone.layer4.0.conv.6.weight", "backbone.layer4.0.conv.7.weight", "backbone.layer4.0.conv.7.bias", "backbone.layer4.0.conv.7.running_mean", "backbone.layer4.0.conv.7.running_var", "backbone.layer4.1.conv.0.weight", "backbone.layer4.1.conv.1.weight", "backbone.layer4.1.conv.1.bias", "backbone.layer4.1.conv.1.running_mean", "backbone.layer4.1.conv.1.running_var", "backbone.layer4.1.conv.3.weight", "backbone.layer4.1.conv.4.weight", "backbone.layer4.1.conv.4.bias", "backbone.layer4.1.conv.4.running_mean", "backbone.layer4.1.conv.4.running_var", "backbone.layer4.1.conv.6.weight", "backbone.layer4.1.conv.7.weight", "backbone.layer4.1.conv.7.bias", "backbone.layer4.1.conv.7.running_mean", "backbone.layer4.1.conv.7.running_var", "backbone.layer4.2.conv.0.weight", "backbone.layer4.2.conv.1.weight", "backbone.layer4.2.conv.1.bias", "backbone.layer4.2.conv.1.running_mean", "backbone.layer4.2.conv.1.running_var", "backbone.layer4.2.conv.3.weight", "backbone.layer4.2.conv.4.weight", "backbone.layer4.2.conv.4.bias", "backbone.layer4.2.conv.4.running_mean", "backbone.layer4.2.conv.4.running_var", "backbone.layer4.2.conv.6.weight", "backbone.layer4.2.conv.7.weight", "backbone.layer4.2.conv.7.bias", "backbone.layer4.2.conv.7.running_mean", "backbone.layer4.2.conv.7.running_var", "backbone.layer4.3.conv.0.weight", "backbone.layer4.3.conv.1.weight", "backbone.layer4.3.conv.1.bias", "backbone.layer4.3.conv.1.running_mean", "backbone.layer4.3.conv.1.running_var", "backbone.layer4.3.conv.3.weight", "backbone.layer4.3.conv.4.weight", "backbone.layer4.3.conv.4.bias", "backbone.layer4.3.conv.4.running_mean", "backbone.layer4.3.conv.4.running_var", "backbone.layer4.3.conv.6.weight", "backbone.layer4.3.conv.7.weight", "backbone.layer4.3.conv.7.bias", "backbone.layer4.3.conv.7.running_mean", "backbone.layer4.3.conv.7.running_var", "backbone.layer5.0.conv.0.weight", "backbone.layer5.0.conv.1.weight", "backbone.layer5.0.conv.1.bias", "backbone.layer5.0.conv.1.running_mean", "backbone.layer5.0.conv.1.running_var", "backbone.layer5.0.conv.3.weight", "backbone.layer5.0.conv.4.weight", "backbone.layer5.0.conv.4.bias", "backbone.layer5.0.conv.4.running_mean", "backbone.layer5.0.conv.4.running_var", "backbone.layer5.0.conv.6.weight", "backbone.layer5.0.conv.7.weight", "backbone.layer5.0.conv.7.bias", "backbone.layer5.0.conv.7.running_mean", "backbone.layer5.0.conv.7.running_var", "backbone.layer5.1.conv.0.weight", "backbone.layer5.1.conv.1.weight", "backbone.layer5.1.conv.1.bias", "backbone.layer5.1.conv.1.running_mean", "backbone.layer5.1.conv.1.running_var", "backbone.layer5.1.conv.3.weight", "backbone.layer5.1.conv.4.weight", "backbone.layer5.1.conv.4.bias", "backbone.layer5.1.conv.4.running_mean", "backbone.layer5.1.conv.4.running_var", "backbone.layer5.1.conv.6.weight", "backbone.layer5.1.conv.7.weight", "backbone.layer5.1.conv.7.bias", "backbone.layer5.1.conv.7.running_mean", "backbone.layer5.1.conv.7.running_var", "backbone.layer5.2.conv.0.weight", "backbone.layer5.2.conv.1.weight", "backbone.layer5.2.conv.1.bias", "backbone.layer5.2.conv.1.running_mean", "backbone.layer5.2.conv.1.running_var", "backbone.layer5.2.conv.3.weight", "backbone.layer5.2.conv.4.weight", "backbone.layer5.2.conv.4.bias", "backbone.layer5.2.conv.4.running_mean", "backbone.layer5.2.conv.4.running_var", "backbone.layer5.2.conv.6.weight", "backbone.layer5.2.conv.7.weight", "backbone.layer5.2.conv.7.bias", "backbone.layer5.2.conv.7.running_mean", "backbone.layer5.2.conv.7.running_var", "backbone.layer6.0.conv.0.weight", "backbone.layer6.0.conv.1.weight", "backbone.layer6.0.conv.1.bias", "backbone.layer6.0.conv.1.running_mean", "backbone.layer6.0.conv.1.running_var", "backbone.layer6.0.conv.3.weight", "backbone.layer6.0.conv.4.weight", "backbone.layer6.0.conv.4.bias", "backbone.layer6.0.conv.4.running_mean", "backbone.layer6.0.conv.4.running_var", "backbone.layer6.0.conv.6.weight", "backbone.layer6.0.conv.7.weight", "backbone.layer6.0.conv.7.bias", "backbone.layer6.0.conv.7.running_mean", "backbone.layer6.0.conv.7.running_var", "backbone.layer6.1.conv.0.weight", "backbone.layer6.1.conv.1.weight", "backbone.layer6.1.conv.1.bias", "backbone.layer6.1.conv.1.running_mean", "backbone.layer6.1.conv.1.running_var", "backbone.layer6.1.conv.3.weight", "backbone.layer6.1.conv.4.weight", "backbone.layer6.1.conv.4.bias", "backbone.layer6.1.conv.4.running_mean", "backbone.layer6.1.conv.4.running_var", "backbone.layer6.1.conv.6.weight", "backbone.layer6.1.conv.7.weight", "backbone.layer6.1.conv.7.bias", "backbone.layer6.1.conv.7.running_mean", "backbone.layer6.1.conv.7.running_var", "backbone.layer6.2.conv.0.weight", "backbone.layer6.2.conv.1.weight", "backbone.layer6.2.conv.1.bias", "backbone.layer6.2.conv.1.running_mean", "backbone.layer6.2.conv.1.running_var", "backbone.layer6.2.conv.3.weight", "backbone.layer6.2.conv.4.weight", "backbone.layer6.2.conv.4.bias", "backbone.layer6.2.conv.4.running_mean", "backbone.layer6.2.conv.4.running_var", "backbone.layer6.2.conv.6.weight", "backbone.layer6.2.conv.7.weight", "backbone.layer6.2.conv.7.bias", "backbone.layer6.2.conv.7.running_mean", "backbone.layer6.2.conv.7.running_var", "backbone.layer7.0.conv.0.weight", "backbone.layer7.0.conv.1.weight", "backbone.layer7.0.conv.1.bias", "backbone.layer7.0.conv.1.running_mean", "backbone.layer7.0.conv.1.running_var", "backbone.layer7.0.conv.3.weight", "backbone.layer7.0.conv.4.weight", "backbone.layer7.0.conv.4.bias", "backbone.layer7.0.conv.4.running_mean", "backbone.layer7.0.conv.4.running_var", "backbone.layer7.0.conv.6.weight", "backbone.layer7.0.conv.7.weight", "backbone.layer7.0.conv.7.bias", "backbone.layer7.0.conv.7.running_mean", "backbone.layer7.0.conv.7.running_var", "neck.downsample2.downsample.0.weight", "neck.downsample2.downsample.1.weight", "neck.downsample2.downsample.1.bias", "neck.downsample2.downsample.1.running_mean", "neck.downsample2.downsample.1.running_var", "neck.downsample3.downsample.0.weight", "neck.downsample3.downsample.1.weight", "neck.downsample3.downsample.1.bias", "neck.downsample3.downsample.1.running_mean", "neck.downsample3.downsample.1.running_var", "neck.downsample4.downsample.0.weight", "neck.downsample4.downsample.1.weight", "neck.downsample4.downsample.1.bias", "neck.downsample4.downsample.1.running_mean", "neck.downsample4.downsample.1.running_var", "rpn_head.rpn2.cls.conv_kernel.0.weight", "rpn_head.rpn2.cls.conv_kernel.1.weight", "rpn_head.rpn2.cls.conv_kernel.1.bias", "rpn_head.rpn2.cls.conv_kernel.1.running_mean", "rpn_head.rpn2.cls.conv_kernel.1.running_var", "rpn_head.rpn2.cls.conv_search.0.weight", "rpn_head.rpn2.cls.conv_search.1.weight", "rpn_head.rpn2.cls.conv_search.1.bias", "rpn_head.rpn2.cls.conv_search.1.running_mean", "rpn_head.rpn2.cls.conv_search.1.running_var", "rpn_head.rpn2.cls.head.0.weight", "rpn_head.rpn2.cls.head.1.weight", "rpn_head.rpn2.cls.head.1.bias", "rpn_head.rpn2.cls.head.1.running_mean", "rpn_head.rpn2.cls.head.1.running_var", "rpn_head.rpn2.cls.head.3.weight", "rpn_head.rpn2.cls.head.3.bias", "rpn_head.rpn2.loc.conv_kernel.0.weight", "rpn_head.rpn2.loc.conv_kernel.1.weight", "rpn_head.rpn2.loc.conv_kernel.1.bias", "rpn_head.rpn2.loc.conv_kernel.1.running_mean", "rpn_head.rpn2.loc.conv_kernel.1.running_var", "rpn_head.rpn2.loc.conv_search.0.weight", "rpn_head.rpn2.loc.conv_search.1.weight", "rpn_head.rpn2.loc.conv_search.1.bias", "rpn_head.rpn2.loc.conv_search.1.running_mean", "rpn_head.rpn2.loc.conv_search.1.running_var", "rpn_head.rpn2.loc.head.0.weight", "rpn_head.rpn2.loc.head.1.weight", "rpn_head.rpn2.loc.head.1.bias", "rpn_head.rpn2.loc.head.1.running_mean", "rpn_head.rpn2.loc.head.1.running_var", "rpn_head.rpn2.loc.head.3.weight", "rpn_head.rpn2.loc.head.3.bias", "rpn_head.rpn3.cls.conv_kernel.0.weight", "rpn_head.rpn3.cls.conv_kernel.1.weight", "rpn_head.rpn3.cls.conv_kernel.1.bias", "rpn_head.rpn3.cls.conv_kernel.1.running_mean", "rpn_head.rpn3.cls.conv_kernel.1.running_var", "rpn_head.rpn3.cls.conv_search.0.weight", "rpn_head.rpn3.cls.conv_search.1.weight", "rpn_head.rpn3.cls.conv_search.1.bias", "rpn_head.rpn3.cls.conv_search.1.running_mean", "rpn_head.rpn3.cls.conv_search.1.running_var", "rpn_head.rpn3.cls.head.0.weight", "rpn_head.rpn3.cls.head.1.weight", "rpn_head.rpn3.cls.head.1.bias", "rpn_head.rpn3.cls.head.1.running_mean", "rpn_head.rpn3.cls.head.1.running_var", "rpn_head.rpn3.cls.head.3.weight", "rpn_head.rpn3.cls.head.3.bias", "rpn_head.rpn3.loc.conv_kernel.0.weight", "rpn_head.rpn3.loc.conv_kernel.1.weight", "rpn_head.rpn3.loc.conv_kernel.1.bias", "rpn_head.rpn3.loc.conv_kernel.1.running_mean", "rpn_head.rpn3.loc.conv_kernel.1.running_var", "rpn_head.rpn3.loc.conv_search.0.weight", "rpn_head.rpn3.loc.conv_search.1.weight", "rpn_head.rpn3.loc.conv_search.1.bias", "rpn_head.rpn3.loc.conv_search.1.running_mean", "rpn_head.rpn3.loc.conv_search.1.running_var", "rpn_head.rpn3.loc.head.0.weight", "rpn_head.rpn3.loc.head.1.weight", "rpn_head.rpn3.loc.head.1.bias", "rpn_head.rpn3.loc.head.1.running_mean", "rpn_head.rpn3.loc.head.1.running_var", "rpn_head.rpn3.loc.head.3.weight", "rpn_head.rpn3.loc.head.3.bias", "rpn_head.rpn4.cls.conv_kernel.0.weight", "rpn_head.rpn4.cls.conv_kernel.1.weight", "rpn_head.rpn4.cls.conv_kernel.1.bias", "rpn_head.rpn4.cls.conv_kernel.1.running_mean", "rpn_head.rpn4.cls.conv_kernel.1.running_var", "rpn_head.rpn4.cls.conv_search.0.weight", "rpn_head.rpn4.cls.conv_search.1.weight", "rpn_head.rpn4.cls.conv_search.1.bias", "rpn_head.rpn4.cls.conv_search.1.running_mean", "rpn_head.rpn4.cls.conv_search.1.running_var", "rpn_head.rpn4.cls.head.0.weight", "rpn_head.rpn4.cls.head.1.weight", "rpn_head.rpn4.cls.head.1.bias", "rpn_head.rpn4.cls.head.1.running_mean", "rpn_head.rpn4.cls.head.1.running_var", "rpn_head.rpn4.cls.head.3.weight", "rpn_head.rpn4.cls.head.3.bias", "rpn_head.rpn4.loc.conv_kernel.0.weight", "rpn_head.rpn4.loc.conv_kernel.1.weight", "rpn_head.rpn4.loc.conv_kernel.1.bias", "rpn_head.rpn4.loc.conv_kernel.1.running_mean", "rpn_head.rpn4.loc.conv_kernel.1.running_var", "rpn_head.rpn4.loc.conv_search.0.weight", "rpn_head.rpn4.loc.conv_search.1.weight", "rpn_head.rpn4.loc.conv_search.1.bias", "rpn_head.rpn4.loc.conv_search.1.running_mean", "rpn_head.rpn4.loc.conv_search.1.running_var", "rpn_head.rpn4.loc.head.0.weight", "rpn_head.rpn4.loc.head.1.weight", "rpn_head.rpn4.loc.head.1.bias", "rpn_head.rpn4.loc.head.1.running_mean", "rpn_head.rpn4.loc.head.1.running_var", "rpn_head.rpn4.loc.head.3.weight", "rpn_head.rpn4.loc.head.3.bias". 
	Unexpected key(s) in state_dict: "epoch", "state_dict", "optimizer".
 
Does anyone have the same problem?
I can training, but I can't demo.
I get this Error info When I executing the demo.py

**Comment by Grundwissen** [comment, marked as answer]
I encountered the same problem.
The pysot framework is saving your weights with DataParallel. By lodading your model in demo.py you have to exclude the keys in state_dict (i.e. in your case "epoch", "state_dict", "optimizer").
I solved this problem by inserting the "load model" code from the test.py to demo.py:
This is a Screenshot from my demo.py file

![Screenshot_20200316_080158](https://user-images.githubusercontent.com/46603930/76730898-7c670580-675c-11ea-95b5-04e642bcf9ed.png)

PS: Don't forget to import the necessary modules (i.e. from pysot.utils.model_load import check_keys, remove_prefix)

**Reply by vic01512** [reply]
nice!!! thank you

**Comment by way790110** [comment]
@Grundwissen 
Thank you for your solution.
I can exexute demo.py.
:smile:

**Comment by Young-eng** [comment]
Same problem, it also works for me, thank you so much @Grundwissen
